packages/contracts/contracts/adapters/UniswapV3Adapter.sol

```diff
@@ -0,0 +1,133 @@
+///         of a single swap and path of a multi-hop swap to ensure no custom-deployed tokens are
+///         used.
+/// @dev This contract is Ownable and non-upgradeable. The only power the owner has is adding
+///      and removing tokens from the whitelist.
+contract UniswapV3Adapter is Ownable {
+    // UniswapV3 SwapRouter
+    ISwapRouter public immutable _swapRouter;
+
+    // Whitelist of tokens that can be used in a path
+    mapping(address => bool) public _allowedTokens;
+
+    /// @notice Event emitted when a token is added to the whitelist
+    event TokenPermissionSet(address token, bool permission);
+
+    // Constructor, sets uniswap swap router contract
+    constructor(address swapRouter) Ownable() {
+        _swapRouter = ISwapRouter(swapRouter);
+    }
+
+    /// @notice Set token permission in whitelist
+    /// @param token Token address
+    /// @param permission Whether token is allowed
+    function setTokenPermission(
+        address token,
+        bool permission
+    ) external onlyOwner {
+        _allowedTokens[token] = permission;
+        emit TokenPermissionSet(token, permission);
+    }
+
+    /// @notice Proxy for exactInputSingle function of Uniswap SwapRouter. Ensures tokenIn
+    ///         is allowed and recipient is caller.
+    /// @param params ExactInputSingleParams (see ISwapRouter)
+    function exactInputSingle(
+        ISwapRouter.ExactInputSingleParams memory params
+    ) external returns (uint256 amountOut) {
+        require(_allowedTokens[params.tokenIn], "!allowed token");
+        require(params.recipient == msg.sender, "!recipient");
+
+        IERC20(params.tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(params.tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInputSingle(params);
+    }
+
+    /// @notice Proxy for exactInput function of Uniswap SwapRouter.
+    ///         Ensures all tokens in path are is allowed and recipient is caller.
+    /// @param params ExactInputParams (see ISwapRouter)
+    function exactInput(
+        ISwapRouter.ExactInputParams memory params
+    ) external returns (uint256 amountOut) {
+        require(tokensAreAllowed(params.path), "!allowed path");
+        require(params.recipient == msg.sender, "!recipient");
+
+        address tokenIn = extractTokenAddressFromPath(params.path, 0);
+
+        IERC20(tokenIn).transferFrom(
+            address(msg.sender),
+            address(this),
+            params.amountIn
+        );
+        IERC20(tokenIn).approve(address(_swapRouter), params.amountIn);
+
+        amountOut = _swapRouter.exactInput(params);
+    }
+
+    /// @notice Checks that all tokens in path are allowed, returns false if any are not
+    /// @param path Path of tokens
+    /// @dev path is a sequence of hops where a hop is (address token1, uint24 poolFee, address
```

Comment:
Oh I see I didn't quite have this right in my comment (I thought the last token was the one without a fee, not the first one, which makes more sense). This is still confusing, I think this is more precisely what we want.

```
/// @dev path is the input token's address, followed by a sequence of hops where a hop is (uint24 poolFee, address tokenAddress) pair. Another way ...
```

Reply:
Rewriting the comment to be more precise:

```
    /// @notice Checks that all tokens in path are allowed, returns false if any are not
    /// @param path Path of tokens
    /// @dev path is logically a sequence of hops where a hop is (address tokenIn, uint24 poolFee, address tokenOut). It is encoded as follows: `(address tokenIn) || hops`, where `hops` is a sequence of `(uint24 fee, address tokenOut)` pairs.
    ///      The length of path will always be `20 + ((20 + 3) * N)` bytes
```